external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala

```diff
@@ -202,4 +204,25 @@ class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationSuite {
     df2.write.jdbc(jdbcUrl, "datescopy", new Properties)
     df3.write.jdbc(jdbcUrl, "stringscopy", new Properties)
   }
+
+  test("SPARK-28151 Test write table with BYTETYPE") {
+    val tableSchema = StructType(Seq(StructField("serialNum", ByteType, true)))
+    val tableData = Seq(Row(10))
+    val df1 = spark.createDataFrame(
+      spark.sparkContext.parallelize(tableData),
+      tableSchema)
+
+    df1.write
+      .format("jdbc")
+      .mode("overwrite")
+      .option("url", jdbcUrl)
+      .option("dbtable", "testTable")
+      .save()
+    val df2 = spark.read
+      .format("jdbc")
+      .option("url", jdbcUrl)
+      .option("dbtable", "byteTable")
+      .load()
+    df2.show()
```

Comment:
Do you want to test anything about the result here?

Reply:
Yes, would add a check of row counts.

Reply:
I'd go ahead and make the test more meaningful; you should test the contents. Also I'd test values besides 10, like negative values and larger than 127.

Reply:
@shivsood if you update the test a bit I think we can add this.

Reply:
@srowen Will update this test soon.  IMO the blocking issue is the exception that i am getting in Catalyst engine ( refer below and how tested section in PR). The fix fails on 3.0 master and branch2-4.  Note : I tested on 2.4.1 and the fix passes. To me looks like some check added between 2.4.1 and now. 

19/08/02 18:25:44 INFO Executor: Finished task 0.0 in stage 7.0 (TID 7). 1197 bytes result sent to driver
19/08/02 18:25:44 INFO TaskSetManager: Finished task 0.0 in stage 7.0 (TID 7) in 43 ms on localhost (executor driver) (1/2)
19/08/02 18:25:44 INFO CodeGenerator: Code generated in 14.586963 ms
19/08/02 18:25:44 ERROR Executor: Exception in task 1.0 in stage 7.0 (TID 8)
java.lang.RuntimeException: Error while encoding: java.lang.RuntimeException: java.lang.Integer is not a valid external type for schema of tinyint
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 0, serialNum), ByteType) AS serialNum#231

Reply:
@srowen @dongjoon-hyun back on this after a while. I have resolved the issue now. The test pass successfully now. Also i have added additional test to 1. test a broader range of ByteType values and also check the col type post read to assert the returned type is ByteType. Test case now pass fully.